### DISCUSSION TITLE
Bluetooth battery status not showing

### DISCUSSION THREAD

**Discussion by Enigma1309** [prasanthrangan/hyprdots#956] [General]
My battery of Bluetooth device was showing on waybar and on dunst notification whenever i connected it but recently it had stopped showing in either of those, it just shows name, it was working till today's morning but now it isn't 
Is this break caused by update, guys please confirm..

**Comment by Enigma1309** [comment]
It's weird it does show sometimes, sometimes it don't now it's showing nvm closing it as solved

**Comment by kRHYME7** [comment]
Hi this is an experimental feature for the waybar. so this behavior is "Normal". What actually happens is the waybar cannot see/detect the string for the battery percentage.Or when connecting a Bluetooth device they don't provide the string for battery percentage.

The stream is in dbus.

Both waybar and dunst are waiting for the stream of string that indicates the battery percentage.In which sometimes it is not present.

**Reply by Enigma1309** [reply]
ohh that's informative thanks 
i've found [this](https://code.whatever.social/questions/49078659/check-battery-level-of-connected-bluetooth-device-on-linux) working i think we can use this with sed and other commands to see battery percentage

**Reply by kRHYME7** [reply]
+ That will be a different implementation/script running in background.  

+ Also, as I have said, Waybar utilizes IPC (Inter-Process Communication) through DBus for its operations (Same as how dunst notify us, for example the battery percentage). It's a hit-and-miss. 
  
+ The suggestions on the link seems like all of them are monitoring/reading the DBus "stream". Would it still be worth it to monitor it on a different script, which likely act the same as the waybar and dunst? 


+ Imagine the stream like this.
 (I use this to watch the battery-notify script for laptops)

![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/acbad234-88e0-4a2f-8185-78bc1e12e872)



+ If you have some free time, try to test ``` upower --dump ``` every time the Bluetooth percentage is NOT working.  If it shows battery percentage, I'll write a script


+ BTW ``` upower ```  reads the dbus also, but it outputs it in a readable way.  


+ Other solutions seem too hacky and may lead to more issues. 

+ The thread always mentioned  "Experimental". This seems an issue in a lower level part of the whole operating system. Like how the device talks to the systems' software. E.g. we can't do much if a device did not report its percentage or if our OS cannot read the report from the device.

**Reply by Enigma1309** [reply]
ohh sorry i wasn't specific enough, i was actually talking about upower --dump, which just works everytime so maybe we can use it to get the battery

**Reply by kRHYME7** [reply]
I see, if ``` upower --dump ``` reports it while waybar cannot read it, maybe it's possible to make a script for it.  I'll test first

**Reply by kRHYME7** [reply]
Updates: 


I tested it but 
 + Waybar and upower detected it
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/b7c88085-38e5-40c2-9bad-7e7b8421f1df)

+ I turned off my Bluetooth device: 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/ec0194ce-5760-4ac7-a29b-2f64b44dcca9)


+ Connected it but this time Bluetooth batt percentage is not present on waybar. 
also upower --dump did not detect it .  
Also, dunst did not detect it. 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/58776035-9665-471a-8bc2-0f2179a97559)
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/cc317934-d117-42ec-a331-5ba2ba48e413)

``` upower ```  Is also using  the dbus too. 

>  BTW upower reads the dbus also, but it outputs it in a readable way.




#### On my end to fix this issues , I just help my Bluetooth device to "ping" to my system. Like clicking buttons,play/pause. Works sometimes.

**Reply by Enigma1309** [reply]
ohh so basically it's a linux issue

**Reply by kRHYME7** [reply]
Yups, somehow like that. We are all almost relying on dbus to communicate packages. "Maybe" this is not an issue on DEs as they have their own set of protocols/ or hax. 

A good example here is this annoying browser bug. 

![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/6628b65b-dbcb-4f11-854d-38dd66187211)

If there is no running file manager(in background), it won't work almost all the time. Reasons are because of different naming, somewhat like that I forgot. We all know that Linux is not that centralized.

**Reply by Enigma1309** [reply]
that might be dolphin related issue, as i have Gnome-files too so my browser opens that instead of dolphin
